code/game/mecha/mecha.dm

```diff
@@ -118,11 +142,240 @@
+			if(decal.glowing)
+				icon_holder.overlays += get_overlay_from_decal(decal, FORCE_OPEN)
+	glow_icon_cache.Insert(getFlatIcon(icon_holder), "[glow_icon]-open")
+	// BROKEN
+	icon_holder.overlays.Cut()
+	icon_holder.icon = create_overlay("[glow_icon]-broken", glow_colour, glow = TRUE, force_state = FORCE_BROKEN)
+	if(icon_decal_root && decals)
+		for(var/datum/mecha/mecha_decal/decal in decals)
+			if(decal.glowing)
+				icon_holder.overlays += get_overlay_from_decal(decal, FORCE_BROKEN)
+	glow_icon_cache.Insert(getFlatIcon(icon_holder), "[glow_icon]-broken")
+
+#define NO_COLOUR null
+#define SUBTRACT_BALANCE "#333333"
+// Takes the data from the supplied decal and passes it into create_overlay, then returns the result as a Mutable Appearance.
+/obj/mecha/proc/get_overlay_from_decal(var/datum/mecha/mecha_decal/decal, var/force_state)
+	var/new_icon_name
+	if(force_state == FORCE_OCCUPIED)
+		if(!decal.has_state_occupied)
+			return
+	else if(force_state == FORCE_OPEN)
+		if(!decal.has_state_open)
+			return
+	else if(force_state == FORCE_BROKEN)
+		if(!decal.has_state_broken)
+			return
+
+	new_icon_name = "[icon_decal_root]-[decal.decal_string][force_state]"
+	var/colour_state = decal.mutable_colour ? decal.decal_colour : NO_COLOUR
+	return create_overlay(new_icon_name, colour_state, decal.glowing, decal.decal_layer, force_state)
+
+
+// Returns a Mutable Appearance based on icon_name to be applied to the base sprite as an overlay.
+/obj/mecha/proc/create_overlay(var/icon_name, var/added_colour = NO_COLOUR, var/glow = FALSE, var/decal_layer = 0, var/force_state)
+	var/icon/I = new(decal_icons, icon_name)
+	if(I)
+		if(added_colour != NO_COLOUR)
+			if(!fast_render_mode)
+				I.Blend(SUBTRACT_BALANCE, ICON_SUBTRACT)
+			I.Blend(added_colour, ICON_ADD)
+			I.Blend(I, ICON_MULTIPLY)
+			if(!fast_render_mode)
+				for(var/datum/mecha/mecha_decal/decal in decals)
+					if(force_state == FORCE_OCCUPIED && !decal.has_state_occupied)
+						continue
+					if(force_state == FORCE_OPEN && !decal.has_state_open)
+						continue
+					if(force_state == FORCE_BROKEN && !decal.has_state_broken)
+						continue
+					if(decal.decal_layer > decal_layer && glow)
+						var/icon/stencil_icon = new(decal_icons, "[icon_decal_root]-[decal.decal_string][force_state]")
+						I = get_icon_difference(I, stencil_icon)
+		var/mutable_appearance/MA = mutable_appearance(I, "")
+		return MA
+	return
+
+#undef FORCE_NONE
+#undef FORCE_OCCUPIED
+#undef FORCE_UNOCCUPIED
+#undef FORCE_BROKEN
+#undef NO_COLOUR
+
+// Add decal to the list and order it according to its layer.
+/obj/mecha/proc/add_decal(var/datum/mecha/mecha_decal/decal)
+	if(!decals)
+		return FALSE
+	for(var/datum/mecha/mecha_decal/checked_decal in decals)
+		if(checked_decal.decal_string == decal.decal_string)
+			checked_decal = decal.clone()
+			return TRUE
+		if(checked_decal.decal_layer > decal.decal_layer)
+			decals.Insert(decals.Find(checked_decal), decal)
+			update_icon(redraw_cache = TRUE)
```

Comment:
I don't see `update_icon` of `/mecha` taking any arguments where it's actually defined - this may compile, but I'd expect it to runtime when this code is run